src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,304 @@
+import { tryFn, hasDebug, readJson, hasVerbose } from './helpers';
+import { shouldTrack } from './shouldTrack';
+import { CompilerBuildResults, Config, PackageJsonData, TaskCommand } from 'src/declarations';
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+import { getCompilerSystem, getCoreCompiler, getStencilCLIConfig } from '../state/stencil-cli-config';
+
+/**
+ * Used as the object sent to the server. Value is the data tracked.
+ */
+export interface Metric {
+  name: string;
+  timestamp: string;
+  source: 'stencil_cli';
+  value: TrackableData;
+}
+
+/**
+ * Used for deeper understanding of how people are using certain output targets
+ **/
+interface OutputTargetOptions {
+  serviceWorker?: boolean;
+}
+
+/**
+ * The task to run in order to collect the duration data point.
+ */
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+/**
+ * The model for the data that's tracked.
+ */
+export interface TrackableData {
+  yarn: boolean;
+  session_id: string;
+  component_count?: number;
+  arguments: string[];
+  targets: string[][] | OutputTargetOptions[][];
+  task: TaskCommand;
+  duration_ms: number;
+  packages: string[];
+  os_name: string;
+  os_version: string;
+  cpu_model: string;
+  typescript: string;
+  rollup: string;
+  system: string;
+  build: string;
+  stencil: string;
+}
+
+/**
+ * Used to within taskBuild to provide the component_count property.
+ * @param result The results of a compiler build.
+ * @returns void
+ */
+export async function telemetryBuildFinishedAction(result: CompilerBuildResults) {
+  const { sys, flags, logger, validatedConfig } = getStencilCLIConfig();
+  const tracking = await shouldTrack(!!flags?.ci);
+
+  if (!tracking) {
+    return;
+  }
+
+  const details = sys.details;
+  const session_id = await getTelemetryToken();
```

Comment:
You may want to consider what happens if the user does something like open their config file and messes with the `tokens.telemetry` field. I think we're assuming that if the config file exists at the moment, `tokens.telemetry` exists and is valid

Reply:
I didn't want to go further than what Ionic's other products do for telemetry. Are you asking me to validate the content of the tokens.telemetry just in case someone has edited the file? I can certainly do that if you would like.

Reply:
How about this, let's make that a follow on story - I think there's some discussion to be had there. IDK if there's any validation on the data type in our data store that would already reject the field if it was malformed. Would you mind adding something to our refinement spreadsheet for next Monday's session?